mitiq/lre/inference/multivariate_richardson.py

```diff
@@ -0,0 +1,199 @@
+        counts = Counter(combo)
+        # Ensure variables are processed in lexicographical order
+        for var in sorted(counts.keys()):
+            count = counts[var]
+            if count > 1:
+                monomial_parts.append(f"{var}**{count}")
+            else:
+                monomial_parts.append(var)
+        monomial = "*".join(monomial_parts)
+        # Handle the case where degree is 0
+        # the tuple in variable_combinations is empty
+        if not monomial:
+            monomial = "1"
+        monomial_basis.append(monomial)
+    # "1" should be the first monomial (degree = 0).
+    # max degree should be the last term
+    return monomial_basis[::-1]
+
+
+def sample_matrix(
+    input_circuit: Circuit,
+    degree: int,
+    fold_multiplier: int,
+    num_chunks: Optional[int] = None,
+) -> NDArray[Any]:
+    r"""
+    Defines the sample matrix required for multivariate extrapolation as
+    defined in :cite:`Russo_2024_LRE`.
+
+    Args:
+        input_circuit: Circuit to be scaled.
+        degree: Degree of the multivariate polynomial.
+        fold_multiplier: Scaling gap required by unitary folding.
+        num_chunks: Number of desired approximately equal chunks. When the
+            number of chunks is the same as the layers in the input circuit,
+            the input circuit is unchanged.
+
+    Returns:
+        Matrix of the evaluated monomial basis terms from the scale factor
+            vectors.
+
+    Raises:
+        ValueError:
+            When the degree for the multinomial is not greater than or
+                equal to 1; when the fold multiplier to scale the circuit is
+                greater than/equal to 1; when the number of chunks for a
+                large circuit is 0 or when the number of chunks in a circuit is
+                greater than the number of layers in the input circuit.
+
+    """
+    if degree < 1:
+        raise ValueError(
+            "Multinomial degree must be greater than or equal to 1."
+        )
+    if fold_multiplier < 1:
+        raise ValueError("Fold multiplier must be greater than or equal to 1.")
+
+    scale_factor_vectors = _get_scale_factor_vectors(
+        input_circuit, degree, fold_multiplier, num_chunks
+    )
+    num_layers = len(scale_factor_vectors[0])
+
+    monomial_terms = full_monomial_basis(num_layers, degree)
+    if len(monomial_terms) != len(scale_factor_vectors):
+        raise ValueError("Sample matrix will not be a square matrix.")
+    sample_matrix = np.zeros((len(monomial_terms), len(monomial_terms)))
+
+    # Evaluate the monomial terms using the values in the scale factor vectors
+    # and insert in the sample matrix
+    # each row is specific to each scale factor vector
+    # each column is a term in the monomial basis
+    for i, point in enumerate(scale_factor_vectors):
+        for j, monomial in enumerate(monomial_terms):
+            var_mapping = {f"λ_{k+1}": point[k] for k in range(num_layers)}
```

Comment:
Creating the `var_mapping` dictionary in each iteration of the nested loop is inefficient. We can move this outside the inner loop.